scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -220,37 +268,91 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
             // (which in case of enum is always true) and replace it with '$'.
             // Now if we have token 'case' and top of stack is '$' we know it is Enum-case.
             // In any other case it is 'match-case' or 'try-case'
-            if (!sepRegions.isEmpty && sepRegions.head == '{') '$' :: sepRegions.tail
-            else '}' :: sepRegions
-          } else if (curr.is[KwEnum]) '{' :: sepRegions
-          else if (curr.is[CaseIntro]) {
-            if (!sepRegions.isEmpty && sepRegions.head == '$') sepRegions
-            else '\u21d2' :: sepRegions
+            if (!sepRegions.isEmpty && sepRegions.head._1 == '{') ('$', 0) :: sepRegions.tail
+            else ('}', 0) :: sepRegions
+          } else if (curr.is[KwEnum]) ('{', 0) :: sepRegions
+          else if (curr.is[CaseIntro] && prev.isNot[KwFor]) {
+            if (!sepRegions.isEmpty && (sepRegions.head._1 == '$' || sepRegions.head._1 == 'E'))
+              sepRegions
+            else ('\u21d2', 0) :: sepRegions
           } else if (curr.is[RightBrace]) {
             var sepRegions1 = sepRegions
-            while (!sepRegions1.isEmpty && (sepRegions1.head != '}' && sepRegions1.head != '$'))
+            while (!sepRegions1.isEmpty && (sepRegions1.head._1 != '}' && sepRegions1.head._1 != '$'))
               sepRegions1 = sepRegions1.tail
             if (!sepRegions1.isEmpty) sepRegions1 = sepRegions1.tail
             sepRegions1
           } else if (curr.is[RightBracket]) {
-            if (!sepRegions.isEmpty && sepRegions.head == ']') sepRegions.tail else sepRegions
+            if (!sepRegions.isEmpty && sepRegions.head._1 == ']') sepRegions.tail else sepRegions
           } else if (curr.is[RightParen]) {
-            if (!sepRegions.isEmpty && sepRegions.head == ')') sepRegions.tail else sepRegions
+            var sepRegions1 = sepRegions
+            while (indentedRegion(sepRegions1)) {
+              parserTokens += new Indentation.Outdent(
+                curr.input,
+                curr.dialect,
+                curr.start,
+                curr.end
+              )
+              parserTokenPositions += currPos
+              sepRegions1 = sepRegions1.tail
+            }
+            if (!sepRegions1.isEmpty && sepRegions1.head._1 == ')') sepRegions1.tail
+            else sepRegions1
           } else if (curr.is[RightArrow]) {
-            if (!sepRegions.isEmpty && sepRegions.head == '\u21d2') sepRegions.tail else sepRegions
+            if (!sepRegions.isEmpty && sepRegions.head._1 == '\u21d2') sepRegions.tail
+            else sepRegions
           } else sepRegions // do nothing for other tokens
         }
+
+        parserTokens += curr
+        parserTokenPositions += currPos
+
+        def isAheadNewLine(): Boolean = {
+          var i = currPos + 1
+          while (i < scannerTokens.length && scannerTokens(i).is[Trivia] && scannerTokens(i)
```

Comment:
opt: implement as `@tailrec def`, won't need to check `.is[LF]` multiple times.

Reply:
changed